### PR TITLE
0213/doodson xo filter

### PR DESCRIPTION
**Pull request by davbyr**
New Doodson_XO_Filter functionality in stats_util, TIDEGAUGE and NEMO.

The basic static method `doodson_xo_filter()` lives in `stats_util.py`. This is a modified version of @jpolton code. Hopefully, I've generalised it a little more to any sized input arrays. This method applies the Doodson weights along a numpy axis then puts Nans in for the first and last 19 elements of the time axis (to remove boundary effects). 

In `NEMO` and `TIDEGAUGE` there is the method `apply_doodson_xo_filter()`, which makes use of the static method `stats_util.doodson_xo_filter()`. This routine is specific to the object and knows which variables/dimensions to call. It assumes that the variable it is given is hourly. The new variable has the same name as the original with the added suffix _dxo. E.g. `ssh_dxo`.

For TIDEGAUGE, there is another new routine called `resample_mean`. This makes use of `xarray.resample` to resample a specified variable to a specified time frequency using averaging (hence the mean). This can be used before applying the doodson_xo_filter. It will create a new variable with the suffix _1H for hourly data (for example). In future, this method should probably be expanded to other resampling methods such as subsampling/interpolation.

Additions:
1. New methods in `TIDEGAUGE`, `NEMO` and `stats_util`.
2. Unit testing updated for `TIDEGAUGE.apply_doodson_xo_filter`
3. Unit testing NOT updated for `NEMO.apply_doodson_xo_filter` as example files dont have enough time elements. I will add as an issue to add this to testing in the future.
4. TIDEGAUGE_tutorial updated. Will also update website page.
5. `TIDEGAUGE.plot_timeseries()` has been updated. You can now supply a list of variables to plot on the same figure. This is demonstrated in the unit_testing and tutorial. However, the qc_colors option has been removed for now (maybe to be reimplemented in future as a general colour variable to the method.)
6. In TIDEGAUGE I have changed the t_dim dimension to time, so that it can be a coordinate dimension. t_dim was causing too many problems since with resample there are potentially multiple time dimensions. This might be good to apply across the board to avoid having to rename dimensions for some methods.

NOTE1: `NEMO.apply_doodson_xo_filter` will result in an entire variable being loaded to memory, which is not fantastic. I have added a note to the docstrin for this method telling the user to subset or ensure enough memory is available.

NOTE2: >=39 Time elements are needed to use `stats_util.doodson_xo_filter()`.

### CODE REVIEW

**Review by anwiseNOCL — APPROVED** [review]
This worked without issue on livljobs. With regarding to renaming _t_dim_ to _time_ I think we should seriously consider it. I can't see a good reason not to have time as a dimensional coordinate.

File: coast/NEMO.py [hunk 1]
```diff
@@ -615,6 +615,26 @@ def differentiate(self, in_varstr, dim='z_dim', out_varstr=None, out_obj=None):
             warn(f"{in_varstr} does not exist in {get_slug(self)} dataset")
             return None
         
+    def apply_doodson_xo_filter(self, var_str):
```
Comment (jpolton) [inline coast/NEMO.py:618]: x0 not xo?

File: coast/NEMO.py [hunk 2]
```diff
@@ -615,6 +615,26 @@ def differentiate(self, in_varstr, dim='z_dim', out_varstr=None, out_obj=None):
...
+        ''' Applies Doodson XO filter to a variable. 
+    
+        Input variable is expected to be hourly.
+        Output is saved back to original dataset as {var_str}_dxo
+        
+        !!WARNING: Will load in entire variable to memory. If dataset large,
+        then subset before using this method or ensure you have enough free 
+        RAM to hold the variable (twice). 
+        
+        DB:: Currently not tested in unit_test.py'''
```
Comment (jpolton) [inline coast/NEMO.py:628]: I think it is now, right?

File: coast/NEMO.py [hunk 3]
```diff
@@ -615,6 +615,26 @@ def differentiate(self, in_varstr, dim='z_dim', out_varstr=None, out_obj=None):
...
+        var = self.dataset[var_str]
+        new_var_str = var_str + '_dxo'
```
Comment (jpolton) [inline coast/NEMO.py:630]: xo or x0. This will pop a number of times so I won't mention it again

File: coast/TIDEGAUGE.py [hunk 4]
```diff
@@ -365,70 +369,58 @@ def plot_on_map_multiple(cls,tidegauge_list, color_var_str = None):
...
+            ax = plt.scatter(x,y, s=10)
             
+        plt.grid()
+        plt.xticks(rotation=45)
+        plt.legend(var_list)
         # Title and axes
         plt.xlabel('Date')
-        plt.ylabel(var_name + ' (m)')
-        plt.title(var_name + ' at site: ' + self.dataset.site_name)
+        plt.ylabel(var_str + ' (m)')
```
Comment (jpolton) [inline coast/TIDEGAUGE.py:422]: It looks like the ylabel will be the last variable in var_list. Is this intended?

**Review by jpolton — APPROVED** [review]
Looks good. A few comments attached.

File: coast/NEMO.py (hunk 3 above)
Comment (davbyr) [reply coast/NEMO.py:630]: x0. I was saying 'Ex Oh' in my head and my fingers typed it literally. Will change.

File: coast/TIDEGAUGE.py (hunk 4 above)
Comment (davbyr) [reply coast/TIDEGAUGE.py:422]: Not any more. Will remove.

File: coast/NEMO.py (hunk 2 above)
Comment (davbyr) [reply coast/NEMO.py:628]: No, NEMO.apply_dooson_xo_filter isn't tested, only TIDEGAUGE.apply_doodson_xo_filter. There isn't enough time steps in the example NEMO data to test it at the moment. I've added this as an issue though.

**Pull request merged**